libs/server/Storage/Functions/MainStore/ReadMethods.cs

```diff
@@ -17,14 +17,44 @@ public bool SingleReader(
             ref SpanByte key, ref RawStringInput input,
             ref SpanByte value, ref SpanByteAndMemory dst, ref ReadInfo readInfo)
         {
-            if (value.MetadataSize != 0 && CheckExpiry(ref value))
+            if (value.MetadataSize == 8 && CheckExpiry(ref value))
             {
                 readInfo.RecordInfo.ClearHasETag();
                 return false;
             }
 
             var cmd = input.header.cmd;
 
+            // Ignore special Vector Set logic if we're scanning, detected with cmd == NONE
+            if (cmd != RespCommand.NONE)
+            {
+                // Vector sets are reachable (key not mangled) and hidden.
+                // So we can use that to detect type mismatches.
+                if (readInfo.RecordInfo.VectorSet && !cmd.IsLegalOnVectorSet())
+                {
+                    // Attempted an illegal op on a VectorSet
+                    readInfo.Action = ReadAction.CancelOperation;
+                    return false;
+                }
+                else if (!readInfo.RecordInfo.VectorSet && cmd.IsLegalOnVectorSet())
+                {
+                    // Attempted a vector set op on a non-VectorSet
+                    readInfo.Action = ReadAction.CancelOperation;
+                    return false;
+                }
+            }
+
+            // GET is used in a number of non-RESP contexts, which messes up existing logic
+            //
+            // Easiest to mark the actually-RESP commands with a < 0 arg1 and roll back to old logic
+            // after the Vector Set checks
+            //
+            // TODO: This is quite hacky, but requires a bunch of non-Vector Set changes - do those and remove
+            if (input.arg1 < 0 && cmd == RespCommand.GET)
```

Comment:
For my learning: I didn't understand this workaround - what's the issue that it's solving?